File: Project/CEC_2008_edge/edgedetectACO.py
import numpy as np
import time

try:
    import matplotlib.pylab as plt
    import matplotlib.image as mpimg
except ImportError:
    pass


def acoedgedetect():
    global v
    plt.close(all)
    clear(all)
    clc
    # image loading
    filename = 'eye'
    img = rgb2gray(mpimg.imread(np.array(np.hstack((filename + '.bmp')))))
    img = np.double(img) / 255.
    [nrow, ncol] = img.shape
    # visiblity function initialization, see equation (4)
    for nMethod in range(1., 5.):
        # Four kernel functions used in the paper, see equations (7)-(10)
        # E: exponential; F: flat; G: gaussian; S:Sine; T:Turkey; W:Wave
        # fprintf('Ant Colony Optimization for Edge detection.\nPlease wait......\n');
        v = np.zeros(img.shape)
        v_norm = 0.
        for rr in range(1., nrow + 1):
            for cc in range(1., ncol + 1):
                # definition of clique
                temp1 = np.array(np.vstack((np.hstack((rr - 2., cc - 1.)), np.hstack((rr - 2., cc + 1.)),
                                            np.hstack((rr - 1., cc - 2.)), np.hstack((rr - 1., cc - 1.)),
                                            np.hstack((rr - 1., cc)), np.hstack((rr - 1., cc + 1.)),
                                            np.hstack((rr - 1., cc + 2.)), np.hstack((rr, cc - 1.)))))
                temp2 = np.array(np.vstack((np.hstack((rr + 2., cc + 1.)), np.hstack((rr + 2., cc - 1.)),
                                            np.hstack((rr + 1., cc + 2.)), np.hstack((rr + 1., cc + 1.)),
                                            np.hstack((rr + 1., cc)), np.hstack((rr + 1., cc - 1.)),
                                            np.hstack((rr + 1., cc - 2.)), np.hstack((rr, cc + 1.)))))
                temp0 = nonzero(np.logical_and(np.logical_and(np.logical_and(np.logical_and(np.logical_and(
                    np.logical_and(np.logical_and(temp1[:, 0] >= 0., temp1[:, 0] < nrow), temp1[:, 1] >= 0.),
                    temp1[:, 1] < ncol), temp2[:, 0] >= 0.), temp2[:, 0] < nrow), temp2[:, 1] >= 0.),
                    temp2[:, 1] < ncol))
                temp11 = temp1[int(temp0) - 1, :]
                temp22 = temp2[int(temp0) - 1, :]
                temp00 = np.zeros(temp11.size)
                for kk in np.arange(1, temp11.size + 1):
                    temp00[int(kk) - 1] = np.abs((
                                                 img[int(temp11[int(kk) - 1, 0]) - 1, int(temp11[int(kk) - 1, 1]) - 1] -
                                                 img[int(temp22[int(kk) - 1, 0]) - 1, int(temp22[int(kk) - 1, 1]) - 1]))

                if temp11.size == 0.:
                    v[int(rr) - 1, int(cc) - 1] = 0.
                    v_norm = v_norm + v[int(rr) - 1, int(cc) - 1]
                else:
                    lambda_val = 1.0
                    _switch_val = nMethod
                    if _switch_val == 1.:
                        # 'F'
                        temp00 *= lambda_val
                    elif _switch_val == 2.:
                        # 'Q'
                        temp00 = lambda_val * temp00 ** 2.
                    elif _switch_val == 3.:
                        # 'S'
                        temp00 = np.sin(np.pi * temp00 / 2. / lambda_val)
                    elif _switch_val == 4.:
                        # 'W'
                        temp00 = np.sin((np.pi * temp00 / lambda_val)) * np.pi * temp00 / lambda_val

            v[int(rr) - 1, int(cc) - 1] = np.sum(np.sum((temp00 ** 2.)))
            v_norm = v_norm + v[int(rr) - 1, int(cc) - 1]
        v /= v_norm
        # do normalization
        v *= 100.
        # pheromone function initialization
        p = 0.0001 * np.ones(img.shape)
        # parameter setting, see Section IV in CEC paper
        alpha = 1.
        # equation (4)
        beta = 0.1
        # equation (4)
        rho = 0.1
        # equation (11)
        phi = 0.05
        # equation (12), i.e., (9) in IEEE-CIM-06
        ant_total_num = 2. * np.round(np.sqrt(np.dot(nrow, ncol)))
        ant_pos_idx = np.zeros(ant_total_num, 2.)
        # record the location of ants
        # initialize the positions of ants
        #     rand('state', sum(clock));
        #     temp = rand(ant_total_num, 2);
        #     ant_pos_idx(:,1) = round(1 + (nrow-1) * temp(:,1)); %row index
        #     ant_pos_idx(:,2) = round(1 + (ncol-1) * temp(:,2)); %column index
        # initialize the position of ants at maximum value of visibility
        # function
        # [~,index]=max(v);
        max_value = max(v)
        index = v.index(max_value)
        ant_pos_idx[0:ncol - 1, 0] = np.arange(0, ncol)
        temp = np.arange(1., ncol + 1)
        temp1 = np.dot(np.ones(0, ncol), ncol)
        ant_pos_idx[int(ncol):, 0] = temp1 - temp
        ant_pos_idx[0:nrow, 1] = index
        ant_pos_idx[int(nrow):, 1] = index
        search_clique_mode = '8'
        # Figure 1
        # define the memory length, the positions in ant's memory are
        # non-admissible positions for the next movement
        if np.dot(nrow, ncol) == 128. * 128.:
            A = 40.
            memory_length = np.round((np.random.rand(1.) * (np.dot(1.15, A) - np.dot(0.85, A)) + np.dot(0.85, A)))
            # memory length
        elif np.dot(nrow, ncol) == 256. * 256.:
            A = 30.
            memory_length = np.round((np.random.rand(1.) * (np.dot(1.15, A) - np.dot(0.85, A)) + np.dot(0.85, A)))
            # memory length

        elif np.dot(nrow, ncol) == 512. * 512.:
            A = 20.
            memory_length = np.round((np.random.rand(1.) * (np.dot(1.15, A) - np.dot(0.85, A)) + np.dot(0.85, A)))
            # memory length

        # record the positions in ant's memory, convert 2D position-index (row, col) into
        # 1D position-index
        ant_memory = np.zeros(ant_total_num, memory_length)
        # System setup
        if np.dot(nrow, ncol) == 128. * 128.:
            total_step_num = 300.
            # the numbe of iterations?
        elif np.dot(nrow, ncol) == 256. * 256.:
            total_step_num = 900.

        elif np.dot(nrow, ncol) == 512. * 512.:
            total_step_num = 1500.

        total_iteration_num = 5.
        for iteration_idx in np.arange(1., total_iteration_num + 1):
            t = time.time()
            # record the positions where ant have reached in the last 'memory_length' iterations
            delta_p = np.zeros(nrow, ncol)
            for step_idx in np.arange(1., total_step_num + 1):
                delta_p_current = np.zeros(nrow, ncol)
                for ant_idx in np.arange(1., ant_total_num + 1):
                    ant_current_row_idx = ant_pos_idx[int(ant_idx) - 1, 0]
                    ant_current_col_idx = ant_pos_idx[int(ant_idx) - 1, 1]  # find the neighborhood of current position
                    if search_clique_mode == '4':
                        rr = ant_current_row_idx
                        cc = ant_current_col_idx
                        ant_search_range_temp = np.array(np.vstack((np.hstack((rr - 1., cc)), np.hstack((rr, cc + 1.)),
                                                                    np.hstack((rr + 1., cc)),
                                                                    np.hstack((rr, cc - 1.)))))
                    elif search_clique_mode == '8':
                        rr = ant_current_row_idx
                        cc = ant_current_col_idx
                        ant_search_range_temp = np.array(np.vstack((np.hstack((rr - 1., cc - 1.)),
                                                                    np.hstack((rr - 1., cc)),
                                                                    np.hstack((rr - 1., cc + 1.)),
                                                                    np.hstack((rr, cc - 1.)), np.hstack((rr, cc + 1.)),
                                                                    np.hstack((rr + 1., cc - 1.)),
                                                                    np.hstack((rr + 1., cc)),
                                                                    np.hstack((rr + 1., cc + 1.)))))

                    # remove the positions our of the image's range
                    temp = nonzero(np.logical_and(np.logical_and(
                        np.logical_and(ant_search_range_temp[:, 0] >= 1., ant_search_range_temp[:, 0] <= nrow),
                        ant_search_range_temp[:, 1] >= 1.), ant_search_range_temp[:, 1] <= ncol))
                    ant_search_range = ant_search_range_temp[int(temp) - 1, :]
                    # calculate the transit prob. to the neighborhood of current
                    # position
                    ant_transit_prob_v = np.zeros(ant_search_range.size, 1.)
                    ant_transit_prob_p = np.zeros(ant_search_range.size, 1.)
                    for kk in np.arange(1., (ant_search_range.size) + 1):
                        temp = np.dot(ant_search_range[int(kk) - 1, 0] - 1., ncol) + ant_search_range[int(kk) - 1, 1]
                        if length(nonzero((ant_memory[int(ant_idx) - 1, :] == temp))) == 0.:
                            # not in ant's memory
                            ant_transit_prob_v[int(kk) - 1] = v[
                                int(ant_search_range[int(kk) - 1, 0]) - 1, int(ant_search_range[int(kk) - 1, 1]) - 1]
                            ant_transit_prob_p[int(kk) - 1] = p[
                                int(ant_search_range[int(kk) - 1, 0]) - 1, int(ant_search_range[int(kk) - 1, 1]) - 1]
                        else:
                            # in ant's memory
                            ant_transit_prob_v[int(kk) - 1] = 0.
                            ant_transit_prob_p[int(kk) - 1] = 0.

                    # if all neighborhood are in memory, then the permissible search range is RE-calculated.
                    if np.logical_or(np.sum(np.sum(ant_transit_prob_v)) == 0.,
                                     np.sum(np.sum(ant_transit_prob_p)) == 0.):
                        for kk in np.arange(1., (ant_search_range.size) + 1):
                            temp = np.dot(ant_search_range[int(kk) - 1, 0] - 1., ncol) + ant_search_range[
                                int(kk) - 1, 1]
                            ant_transit_prob_v[int(kk) - 1] = v[
                                int(ant_search_range[int(kk) - 1, 0]) - 1, int(ant_search_range[int(kk) - 1, 1]) - 1]
                            ant_transit_prob_p[int(kk) - 1] = p[
                                int(ant_search_range[int(kk) - 1, 0]) - 1, int(ant_search_range[int(kk) - 1, 1]) - 1]

                    ant_transit_prob = ant_transit_prob_v ** alpha * ant_transit_prob_p ** beta / np.sum(
                        np.sum((ant_transit_prob_v ** alpha * ant_transit_prob_p ** beta)))
                    # generate a random number to determine the next position.
                    random.set_state(np.sum((100. * clock)))
                    temp = nonzero((np.cumsum(ant_transit_prob) >= np.random.rand(1.)), 1.)
                    ant_next_row_idx = ant_search_range[int(temp) - 1, 0]
                    ant_next_col_idx = ant_search_range[int(temp) - 1, 1]
                    if length(ant_next_row_idx) == 0.:
                        ant_next_row_idx = ant_current_row_idx
                        ant_next_col_idx = ant_current_col_idx

                    ant_pos_idx[int(ant_idx) - 1, 0] = ant_next_row_idx
                    ant_pos_idx[int(ant_idx) - 1, 1] = ant_next_col_idx
                    # record the delta_p_current
                    delta_p_current[
                        int(ant_pos_idx[int(ant_idx) - 1, 0]) - 1, int(ant_pos_idx[int(ant_idx) - 1, 1]) - 1] = 1.
                    # record the new position into ant's memory
                    if step_idx <= memory_length:
                        ant_memory[int(ant_idx) - 1, int(step_idx) - 1] = np.dot(ant_pos_idx[int(ant_idx) - 1, 0] - 1.,
                                                                                 ncol) + ant_pos_idx[
                                                                              int(ant_idx) - 1, 1]
                    elif step_idx > memory_length:
                        ant_memory[int(ant_idx) - 1, :] = np.roll(ant_memory[int(ant_idx) - 1, :],
                                                                  np.array(np.hstack((0., -1.))))
                        ant_memory[int(ant_idx) - 1, int(0) - 1] = np.dot(ant_pos_idx[int(ant_idx) - 1, 0] - 1., ncol) + \
                                                                   ant_pos_idx[int(ant_idx) - 1, 1]

                    # update the pheromone function (10) in IEEE-CIM-06
                    p = ((1. - rho) * p + rho * delta_p_current * v) * delta_p_current + p * np.abs(
                        (1. - delta_p_current))

                # end of ant_idx
                # update the pheromone function see equation (9) in IEEE-CIM-06
                delta_p = delta_p + (delta_p_current > 0.) > 0.
                p *= (1. - phi)
                # equation (9) in IEEE-CIM-06
            # end of step_idx

            # create the image at every iteration to see the difference
            T = func_seperate_two_class(p)
            # eq. (13)-(21), Calculate the threshold to seperate the edge map into two class
            mpimg.imsave(np.array(np.hstack((filename, nMethod, iteration_idx))),
                         np.uint8(np.abs(((p >= T) * 255. - 255.))), plt.gray(256.), 'bmp')
            timespent = time.time() - t
            print 'time spent for ' + iteration_idx + ' iteration: ' + timespent + '\n'
            # end of iteration_idx
            # generate edge map matrix
            # It uses pheromone function to determine edge?
        print 'Done!\n'
    # end of nMethod
    return


def func_seperate_two_class(I):
    # Local Variables: MAT, mu3, level, I, Threshold, mu2, N, mu, i, T, MBT, counts
    # Function calls: hist, sum, abs, cumsum, func_seperate_two_class
    #   ISODATA Compute global image threshold using iterative isodata method.
    #   LEVEL = ISODATA(I) computes a global threshold (LEVEL) that can be
    #   used to convert an intensity image to a binary image with IM2BW. LEVEL
    #   is a normalized intensity value that lies in the range [0, 1].
    #   This iterative technique for choosing a threshold was developed by Ridler and Calvard .
    #   The histogram is initially segmented into two parts using a starting threshold value such as 0 = 2B-1,
    #   half the maximum dynamic range.
    #   The sample mean (mf,0) of the gray values associated with the foreground pixels and the sample mean (mb,0)
    #   of the gray values associated with the background pixels are computed. A new threshold value 1 is now computed
    #   as the average of these two sample means. The process is repeated, based upon the new threshold,
    #   until the threshold value does not change any more.
    #
    # Reference :T.W. Ridler, S. Calvard, Picture thresholding using an iterative selection method,
    #            IEEE Trans. System, Man and Cybernetics, SMC-8 (1978) 630-632.
    # Convert all N-D arrays into a single column.  Convert to uint8 for
    # fastest histogram computation.
    I = I.flatten(1)
    # STEP 1: Compute mean intensity of image from histogram, set T=mean(I)
    [counts, N] = plt.hist(I, 256.)
    i = 1.
    mu = np.cumsum(counts)
    T[int(i) - 1] = matdiv(np.sum((N * counts)), mu[int(0) - 1])
    # STEP 2: compute Mean above T (MAT) and Mean below T (MBT) using T from
    # step 1
    mu2 = np.cumsum(counts[int((N <= T[int(i) - 1])) - 1])
    MBT = matdiv(np.sum((N[int((N <= T[int(i) - 1])) - 1] * counts[int((N <= T[int(i) - 1])) - 1])), mu2[int(0) - 1])
    mu3 = np.cumsum(counts[int((N > T[int(i) - 1])) - 1])
    MAT = matdiv(np.sum((N[int((N > T[int(i) - 1])) - 1] * counts[int((N > T[int(i) - 1])) - 1])), mu3[int(0) - 1])
    i = i + 1.
    T[int(i) - 1] = (MAT + MBT) / 2.
    # STEP 3 to n: repeat step 2 if T(i)~=T(i-1)
    Threshold = T[int(i) - 1]
    while np.abs((T[int(i) - 1] - T[int((i - 1.)) - 1])) >= 1.:
        mu2 = np.cumsum(counts[int((N <= T[int(i) - 1])) - 1])
        MBT = matdiv(np.sum((N[int((N <= T[int(i) - 1])) - 1] * counts[int((N <= T[int(i) - 1])) - 1])),
                     mu2[int(0) - 1])
        mu3 = np.cumsum(counts[int((N > T[int(i) - 1])) - 1])
        MAT = matdiv(np.sum((N[int((N > T[int(i) - 1])) - 1] * counts[int((N > T[int(i) - 1])) - 1])), mu3[int(0) - 1])
        i = i + 1.
        T[int(i) - 1] = (MAT + MBT) / 2.
        Threshold = T[int(i) - 1]

    # Normalize the threshold to the range [i, 1].
    level = Threshold
    return [level]

File: Project/edgedetectACO.m
function edgedetectACO
%
% lekhak: varsh007@umn.edu
%
% Input:
% gray image with a square size
%
global eta visibility lambda pheromone alpha beta rho phi steps_M iterations L
global ants_K ant_movement T
close all;
clear;
clc;

%parameter setting, 
alpha = 1;      
beta = 0.1; 
rho = 0.1;  
phi = 0.01; 
iterations = 10;
L = 100;
% image loading
filename = 'lenac';
%for reading the RGB images 
img=rgb2gray(imread([filename '.png']));
%for reading bw images
% img=imread([filename '.png']);
img = double(img)./255;
[rows, cols] = size(img);
%initialization
% we have K ants in total, equal to the the width of the image, due to an assumption that each
% column will have one ant.
ants_K = cols;
ant_movement = zeros(ants_K, L);
ants_position_coordinates = zeros(ants_K, 2); 
% pheromone matrix initialization
pheromone = 0.0001 .* ones(size(img));
% initializing it here due to the fact of second update rule
pheromone_init=pheromone;
%visiblity function initialization
eta = zeros(size(img));
for X =1:rows
    for Y=1:cols        
        vis1 = [X-2 Y-1; X-2 Y+1; X-1 Y-2; X-1 Y-1; X-1 Y; X-1 Y+1; X-1 Y+2; X Y-1];
        vis2 = [X+2 Y+1; X+2 Y-1; X+1 Y+2; X+1 Y+1; X+1 Y; X+1 Y-1; X+1 Y-2; X Y+1];
        vis = find(vis1(:,1)>=1 & vis1(:,1)<=rows & vis1(:,2)>=1 & vis1(:,2)<=cols & vis2(:,1)>=1 & vis2(:,1)<=rows & vis2(:,2)>=1 & vis2(:,2)<=cols);        
        vis1_in_range = vis1(vis, :);
        vis2_in_range = vis2(vis, :);
        visibility = zeros(size(vis1_in_range,1),1);
        for xy = 1:size(vis1_in_range,1)
            % modulus of visibility function values, 8 values for 8
            % connectivity neighborhood
            visibility(xy) = abs(img(vis1_in_range(xy,1), vis1_in_range(xy,2))-img(vis2_in_range(xy,1), vis2_in_range(xy,2)));
        end
        
        if size(vis1_in_range,1) == 0
            eta(X, Y) = 0; 
        else
            lambda = 1;
            visibility = sin(pi .* visibility./2./lambda);
            % l2 norm of the function f(x) = sin(pi*(x)/2lambda)
            % l1 norm of the function f(x) 
            % find the maximum of the I differences and then multiply it by variance of the neighborhood
            % do most manipulations here to find the best heuristic 
            eta(X, Y) = (std(visibility)^2)*max(visibility);
        end
    end
end

eta = eta./max(max(eta));  %do normalization dividing by max value of the overall matrix

%initialize the position of ants at maximum value of visibility
%function
[~,index]=max(eta);
%initialize the position of ants at minimum value of visibility
% [~,index]=min(eta);
%initialize the position of ants at mean value of visibility function 
% meanValue=mean(eta,1);
% [~,index]=min(abs(eta-repmat(meanValue,rows,1)));
ants_position_coordinates(1:cols,2)= 1:cols;
% for twice the sqrt(M1xM2) ants
%     in_range_values=0:cols-1;
%     vis1=ones(1,cols)*cols;
%     ants_position_coordinates(cols+1:end,1)=vis1-in_range_values;
ants_position_coordinates(1:rows,1)=index;
%     ants_position_coordinates(rows+1:end,2)=index;

for iterator = 1: iterations
    tic;
    delta_p = zeros(rows, cols);  
    for step_iterator = 1: L
        delta_p_current = zeros(rows, cols);
        for ant_iterator = 1:ants_K
            Kth_ant_row = ants_position_coordinates(ant_iterator,1);
            Kth_ant_col = ants_position_coordinates(ant_iterator,2);
            % how to move from I,J to K,L - Neighborhood move
            X = Kth_ant_row;
            Y = Kth_ant_col;
            neighborhood = [X-1 Y-1; X-1 Y; X-1 Y+1; X Y-1; X Y+1; X+1 Y-1; X+1 Y; X+1 Y+1];
            ant_vis = neighborhood(neighborhood(:,1)>=1 & neighborhood(:,1)<=rows & neighborhood(:,2)>=1 & neighborhood(:,2)<=cols,:);
            %calculate the transit probability function as defined in the
            %paper
            ant_visibility_trp = zeros(size(ant_vis,1),1);
            ant_pheromone_trp = zeros(size(ant_vis,1),1);
            for xy = 1:size(ant_vis,1)
                value = (ant_vis(xy,1)-1)*rows + (ant_vis(xy,2));                
                if isempty(ant_movement(ant_movement(ant_iterator,:)==value))
                    ant_visibility_trp(xy) = eta(ant_vis(xy,1), ant_vis(xy,2));
                    ant_pheromone_trp(xy) = pheromone(ant_vis(xy,1), ant_vis(xy,2));
                end
            end
            if (sum(sum(ant_visibility_trp))==0) || (sum(sum(ant_pheromone_trp))==0)
                for xy = 1:size(ant_vis,1)
                    ant_visibility_trp(xy) = eta(ant_vis(xy,1), ant_vis(xy,2));
                    ant_pheromone_trp(xy) = pheromone(ant_vis(xy,1), ant_vis(xy,2));
                end
            end
            trp = (ant_visibility_trp.^alpha) .* (ant_pheromone_trp.^beta) ./ (sum(sum((ant_visibility_trp.^alpha) .* (ant_pheromone_trp.^beta))));
            
            value = find(cumsum(trp)>=rand(1),1);
            ant_next_row_idx = ant_vis(value,1);
            ant_next_col_idx = ant_vis(value,2);
            if isempty(length(ant_next_row_idx))
                ant_next_row_idx = Kth_ant_row;
                ant_next_col_idx = Kth_ant_col;
            end
            ants_position_coordinates(ant_iterator,1) = ant_next_row_idx;
            ants_position_coordinates(ant_iterator,2) = ant_next_col_idx;
            %record the delta_p_current
            delta_p_current(ants_position_coordinates(ant_iterator,1), ants_position_coordinates(ant_iterator,2)) = 1;
            % record the new position into ant's memory
            ant_movement(ant_iterator,step_iterator) = (ants_position_coordinates(ant_iterator,1)-1)*cols + ants_position_coordinates(ant_iterator,2);
        % local update     
        % first part if i,j belongs to current ant 
        % second part otherwise 
        pheromone = ((1-rho).*pheromone + rho.*eta.*delta_p_current).*delta_p_current + pheromone.*(abs(1-delta_p_current));
        end % end of ant_iterator
        % update the pheromone function globally       
        delta_p = (delta_p + (delta_p_current>0))>0;
        pheromone = ((1-phi).*pheromone.*delta_p + phi.*pheromone_init);
    end % end of step_iterator
    % create the image at every iteration to see the difference
    fprintf('pheromone sum of iteration  %d: %2f\n', iterator, sum(sum(pheromone)));
    fprintf('deltaP sum of iteration %d: %2f\n', iterator, sum(sum(delta_p)));
    T = classifier(pheromone); 
    imwrite(uint8(abs((pheromone>=T).*255-255)), gray(256), [filename '_edge_aco_' num2str(iterator) '.bmp'], 'bmp');
    timespent=toc;
    fprintf('timespent for %d iteration: %2f\n', iterator, timespent);
end % end of iterator
